modules/core/src/main/java/org/apache/ignite/internal/processors/datastructures/DataStructuresProcessor.java

```diff
@@ -911,18 +912,28 @@ private CacheConfiguration metaCacheConfiguration(CollectionConfiguration cfg, S
 
     /**
      * @param cfg Collection configuration.
-     * @return Cache name.
      * @param grpName Group name.
+     * @param shared Shared mode (single cache for multiple data structure instances).
+     * @param dsName Datastructure name.
+     * @return Cache name.
      * @throws IgniteCheckedException If failed.
      */
-    @Nullable private IgniteInternalCache compatibleCache(CollectionConfiguration cfg, String grpName)
-        throws IgniteCheckedException
-    {
+    @Nullable private IgniteInternalCache compatibleCache(CollectionConfiguration cfg, String grpName,
+        boolean shared, String dsName)
+        throws IgniteCheckedException {
+
         String cacheName = DS_CACHE_NAME_PREFIX + cfg.getAtomicityMode() + "_" + cfg.getCacheMode() + "_" +
             cfg.getBackups() + "@" + grpName;
 
         IgniteInternalCache cache = ctx.cache().cache(cacheName);
 
+        if (!shared && (cache == null || cache.get(new GridCacheSetHeaderKey(dsName)) == null)) {
```

Comment:
scenario : in cluster with old and new nodes created old set. Old nodes left grid, and now we want to get old cache from new nodes.